### PR TITLE
Add integration test for the Platform Services app

### PR DESCRIPTION
**Pull request by szakarias**

### CODE REVIEW

File: dev/devicelab/bin/tasks/platform_services_ios__start_up.dart [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+import 'dart:async';
+
+import 'package:flutter_devicelab/tasks/perf_tests.dart';
+import 'package:flutter_devicelab/framework/adb.dart';
+import 'package:flutter_devicelab/framework/framework.dart';
+
+Future<Null> main() async {
+  deviceOperatingSystem = DeviceOperatingSystem.ios;
+  await task(createFlutterGalleryStartupTest());
```
Comment (yjbanov) [inline dev/devicelab/bin/tasks/platform_services_ios__start_up.dart]: Did you mean `createPlatformServicesStartupTest()`?

File: dev/devicelab/bin/tasks/platform_services__build.dart [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:flutter_devicelab/tasks/perf_tests.dart';
+import 'package:flutter_devicelab/framework/framework.dart';
+
+Future<Null> main() async {
+  await task(createPlatformServicesBuildTest());
```
Comment (yjbanov) [inline dev/devicelab/bin/tasks/platform_services__build.dart]: You don't need a build test unless you are interested in code size and build speed metrics. The start-up test has to build anyway and therefore will make sure that we can build the example app successfully.

File: dev/devicelab/lib/tasks/perf_tests.dart [hunk 3]
```diff
@@ -67,6 +67,16 @@ TaskFunction createGalleryBackButtonMemoryTest() {
   );
 }
 
+TaskFunction createPlatformServicesStartupTest() {
+  return new StartupTest(
+      '${flutterDirectory.path}/examples/platform_services',
```
Comment (yjbanov) [inline dev/devicelab/lib/tasks/perf_tests.dart]: My understanding was that we are interested in testing correctness of platform services. If so, then `StartupTest` may not be a suitable way of testing them. Consider writing a driver test that exercises the platform services, and then a devicelab task that runs the driver test.

**Review by yjbanov — COMMENTED** [review]
You've got all the devicelab bits right. I'm not really sure that `StartupTest` is the right thing to do if our goal is to test correctness. A start-up test waits for the first frame of the application to render, then quits immediately declaring a victory. A driver test might be a better option, since it would allow you to exercise the platform services functionality.

**Comment by szakarias** [comment]
Actually my initial thought was just to check if the app could build and launch, but you are right of course, that testing functionality is much better! I haven't cleaned up/removed the old tests yet, but I would be very happy if you could take a look at the driver test and let me know if this is the way forward.

File: examples/platform_services/test_driver/button_tap_test.dart [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+        expect(batteryLevelLabel, isNotNull);
+
+        SerializableFinder button = find.text('Get Battery Level');
+        await driver.waitFor(button);
+        await driver.tap(button);
+
+      expect(
+        retry(
+          () async {
+            return await driver.getText(batteryLevelLabel);
```
Comment (yjbanov) [inline examples/platform_services/test_driver/button_tap_test.dart]: I think the retry API is too heavyweight for this test. How about this:

```dart
String batteryLevel;
while(batteryLevel == null || batteryLevel.isEmpty) {
  batteryLevel = await driver.getText(batteryLevelLabel);
}
expect(batteryLevel, isNotEmpty);
```

The test framework should (hopefully) handle the timeout if the text never appears.

File: dev/devicelab/bin/tasks/platform_services__build.dart [hunk 5]
```diff
@@ -0,0 +1,12 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
```
Comment (yjbanov) [inline dev/devicelab/bin/tasks/platform_services__build.dart]: Each task defined in `bin/tasks` creates a new column in our [test matrix](https://flutter-dashboard.appspot.com/build.html). We therefore try to make our tests coarse grained. If we want to test multiple things, we bundle them in one task. For example, `flutter_gallery__transition_perf` tests all demo apps in the gallery app and measures performance all as part of one task. I intend to fix this at some point by supporting "task bundles". For now, I think we just need one task for platform services that tests everything we need. So let's remove `platform_services__build.dart` and do everything we need in the other one.

File: dev/devicelab/bin/tasks/platform_services__button_tap.dart [hunk 6]
```diff
@@ -0,0 +1,14 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
```
Comment (yjbanov) [inline dev/devicelab/bin/tasks/platform_services__button_tap.dart]: `button_tap` is too much detail for a task. We want the task to be coarse-grained. Let's call it `platform_services_test` and encapsulate the specifics.

File: dev/devicelab/bin/tasks/platform_services__start_up.dart [hunk 7]
```diff
@@ -0,0 +1,14 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
```
Comment (yjbanov) [inline dev/devicelab/bin/tasks/platform_services__start_up.dart]: Let's remove this task. One task should be sufficient to test platform services. We can on-board more fine-grained tests after we teach our test infrastructure to scale in the future.

File: dev/devicelab/bin/tasks/platform_services_ios__button_tap.dart [hunk 8]
```diff
@@ -0,0 +1,14 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
```
Comment (yjbanov) [inline dev/devicelab/bin/tasks/platform_services_ios__button_tap.dart]: Similarly, let's rename this to `platform_services_test_ios`.

File: dev/devicelab/bin/tasks/platform_services_ios__start_up.dart [hunk 9]
```diff
@@ -0,0 +1,14 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
```
Comment (yjbanov) [inline dev/devicelab/bin/tasks/platform_services_ios__start_up.dart]: Let's remove this task.

File: dev/devicelab/lib/tasks/perf_tests.dart [hunk 10]
```diff
@@ -9,6 +9,15 @@ import '../framework/adb.dart';
 import '../framework/framework.dart';
 import '../framework/utils.dart';
 
+
+TaskFunction createPlatformServiceDriverTest() {
+  return new PerfTest(
```
Comment (yjbanov) [inline dev/devicelab/lib/tasks/perf_tests.dart]: Could you please create a copy of `PerfTest` called `DriverTest` and remove the performance metrics? We don't want to publish metrics that we do not set goals and baselines for.

File: dev/devicelab/manifest.yaml [hunk 11]
```diff
@@ -74,6 +74,13 @@ tasks:
     stage: devicelab
     required_agent_capabilities: ["has-android-device"]
 
+  platform_services__build:
```
Comment (yjbanov) [inline dev/devicelab/manifest.yaml]: let's remove

File: dev/devicelab/manifest.yaml [hunk 12]
```diff
@@ -83,6 +90,13 @@ tasks:
     stage: devicelab
     required_agent_capabilities: ["has-android-device"]
 
+  platform_services__button_tap:
```
Comment (yjbanov) [inline dev/devicelab/manifest.yaml]: let's rename to `platform_services_test`

File: dev/devicelab/manifest.yaml [hunk 13]
```diff
@@ -138,9 +152,21 @@ tasks:
     stage: devicelab
     required_agent_capabilities: ["has-android-device"]
 
+  platform_services__start_up:
```
Comment (yjbanov) [inline dev/devicelab/manifest.yaml]: let's remove

File: dev/devicelab/manifest.yaml [hunk 14]
```diff
@@ -138,9 +152,21 @@ tasks:
...
+    description: >
+      Measures the startup time of the Platform Services app on Android.
+    stage: devicelab
+    required_agent_capabilities: ["has-android-device"]
 
   # iOS on-device tests
 
+  platform_services_ios__button_tap:
```
Comment (yjbanov) [inline dev/devicelab/manifest.yaml]: let's rename to `platform_services_test_ios`

File: dev/devicelab/manifest.yaml [hunk 15]
```diff
@@ -179,6 +205,12 @@ tasks:
     stage: devicelab_ios
     required_agent_capabilities: ["has-ios-device"]
 
+  platform_services_ios__start_up:
```
Comment (yjbanov) [inline dev/devicelab/manifest.yaml]: let's remove

**Review by yjbanov — CHANGES_REQUESTED** [review]
Getting very close. Let's restructure the tasks a little and it should be good to go.

**Comment by szakarias** [comment]
Thank you so much for the comments, would you please take another look?

**Review by yjbanov — APPROVED** [review]
LGTM! This marks the day we finally began testing platform services! Hurrah! :)

**Comment by Hixie** [comment]
Unfortunately the tests are red on iOS.

**Comment by yjbanov** [comment]
> Unfortunately the tests are red on iOS.

That was half-expected.

**Comment by szakarias** [comment]
Fixed with #8946 . It seems an old version of Xcode is used for the tests. Would it be possible to upgrade to Xcode 8?

**Comment by yjbanov** [comment]
> Fixed with #8946

Thanks!

> Would it be possible to upgrade to Xcode 8?

It would take a little work, because Xcode 8 has new provisioning profile rules that our test infra needs to learn.

**Comment by szakarias** [comment]
I see. Unfortunately, Xcode will automatically undo the change. That is, every time a storyboard file is opened it is saved in the new format.

**Pull request merged**